controlplane/pkg/sidecars/nsm-monitor.go

```diff
@@ -0,0 +1,238 @@
+					continue
+				} else {
+					recovery = true
+				}
+			}
+			select {
+			case err = <-monitorClient.ErrorChannel():
+				logrus.Fatalf(nsmMonitorLogWithParamFormat, "NSM die, re-connecting", err)
+				for _, c := range c.connections {
+					c.State = connection.State_DOWN // Mark all as down.
+				}
+				continue
+			case event := <-monitorClient.EventChannel():
+				if event.EventType() == monitor.EventTypeInitialStateTransfer {
+					logrus.Infof(nsmMonitorLogFormat, "Monitor started")
+					initRecieved = true
+				}
+
+				for _, entity := range event.Entities() {
+					switch event.EventType() {
+					case monitor.EventTypeInitialStateTransfer, monitor.EventTypeUpdate:
+						c.updateConnection(entity)
+					case monitor.EventTypeDelete:
+						logrus.Infof(nsmMonitorLogFormat, "Connection closed")
+						return
+					}
+				}
+			case <-c.stop:
+				if c.helper != nil {
+					c.helper.Stopped()
+					logrus.Infof(nsmMonitorLogFormat, "Processing stop")
+					return
+				}
+			}
+		}
+	}
+}
+
+func (c *nsmMonitorApp) updateConnection(entity monitor.Entity) {
+	conn, ok := entity.(*connection.Connection)
+	// update connections
+	if ok {
+		if existingConn, exists := c.connections[conn.Id]; exists {
+			logrus.Infof(nsmMonitorLogWithParamFormat, "Connection updated", fmt.Sprintf("%v %v", existingConn, conn))
+		} else {
+			logrus.Infof(nsmMonitorLogWithParamFormat, "Initial connection accepted", conn)
+		}
+		c.connections[conn.Id] = conn
+	}
+}
+
+func (c *nsmMonitorApp) waitRetry() {
+	logrus.Errorf(nsmMonitorLogWithParamFormat, "Retry delay %v sec", nsmMonitorRetryDelay)
+	<-time.After(nsmMonitorRetryDelay * time.Second)
+}
+
+func (c *nsmMonitorApp) performRecovery(nsmClient *client.NsmClient) bool {
+	logrus.Infof(nsmMonitorLogFormat, "Performing recovery if needed...")
+
+	needRetry := false
+	for _, conn := range c.connections {
+		if conn.State == connection.State_UP {
+			continue
+		}
+		cClone := (conn.Clone()).(*connection.Connection)
+
+		ipCtx := cClone.Context.IpContext
+		if ipCtx != nil {
+			if ipCtx.DstIpAddr != "" {
+				ipCtx.DstIpRequired = true
+			}
+			if ipCtx.SrcIpAddr != "" {
+				ipCtx.SrcIpRequired = true
+			}
+		}
+
+		outgoingRequest := networkservice.NetworkServiceRequest{
+			Connection: cClone,
+			MechanismPreferences: []*connection.Mechanism{
+				conn.Mechanism,
+			},
+		}
+		if c.helper != nil {
+			c.helper.Healing(cClone)
+		}
+
+		newConn, err := nsmClient.PerformRequest(&outgoingRequest)
+		if err != nil {
+			logrus.Errorf(nsmMonitorLogWithParamFormat, "failed to restore connection. Will retry", err)
+			// Let's drop connection id, since we failed one time.
+			conn.Id = "-"
+			needRetry = true
+			continue
+		} else {
+			logrus.Errorf(nsmMonitorLogWithParamFormat, "connection restored", newConn)
+			delete(c.connections, conn.Id)
+			c.connections[newConn.Id] = newConn
+		}
+		if c.helper != nil {
+			c.helper.ProcessHealing(newConn, err)
+		}
+	}
+	if needRetry {
+		c.waitRetry()
```

Comment:
`performRecovery` shouldn't call `c.waitRetry()` it is not its responsibility to make delays between recovery. Caller of `performRecovery` has to do it

Reply:
thanks, fixed